flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/CoordinatorEventsToStreamOperatorRecipientExactlyOnceITCase.java

```diff
@@ -0,0 +1,549 @@
+
+    @Test
+    public void testUnalignedCheckpointWithSubtaskFailure() throws Exception {
+        env.getCheckpointConfig().enableUnalignedCheckpoints();
+        executeAndVerifyResults(
+                env,
+                new EventReceivingOperatorWithFailureFactory<>(
+                        "eventReceivingWithFailure", NUM_EVENTS, DELAY));
+        assertThat(
+                        TestScript.getForOperator("eventReceivingWithFailure-subtask0")
+                                .hasAlreadyFailed())
+                .isTrue();
+    }
+
+    private void executeAndVerifyResults(
+            StreamExecutionEnvironment env, EventReceivingOperatorFactory<Long, Long> factory)
+            throws Exception {
+        // The event receiving operator is not chained together with the source operator, so that
+        // when checkpoint barriers are injected into sources, the event receiving operator has not
+        // started checkpoint yet.
+        env.addSource(new ManuallyClosedSourceFunction<>(), TypeInformation.of(Long.class))
+                .transform(
+                        "blockCheckpointBarrier",
+                        TypeInformation.of(Long.class),
+                        new BlockCheckpointBarrierOperator<>())
+                .disableChaining()
+                .transform(factory.name, TypeInformation.of(Long.class), factory)
+                .addSink(new DiscardingSink<>());
+
+        JobExecutionResult executionResult =
+                MINI_CLUSTER
+                        .getMiniCluster()
+                        .executeJobBlocking(env.getStreamGraph().getJobGraph());
+
+        List<Integer> receivedInts =
+                executionResult.getAccumulatorResult(EventReceivingOperator.ACCUMULATOR_NAME);
+        checkListContainsSequence(receivedInts, NUM_EVENTS);
+    }
+
+    /** A mock source function that does not collect any stream record and closes on demand. */
+    private static class ManuallyClosedSourceFunction<T> implements SourceFunction<T> {
+
+        /** Whether the source function should be closed to finish the job. */
+        private static boolean isCloseSource;
+
+        @Override
+        public void run(SourceContext<T> ctx) throws Exception {
+            while (!isCloseSource) {
+                Thread.sleep(100);
+            }
+        }
+
+        @Override
+        public void cancel() {}
+    }
+
+    /**
+     * A stream operator that blocks the checkpoint barrier until the coordinator has sent events to
+     * its subtask. It helps to guarantee that there are events being sent when the coordinator has
+     * completed the first checkpoint while the subtask has not yet.
+     */
+    private static class BlockCheckpointBarrierOperator<T> extends AbstractStreamOperator<T>
+            implements OneInputStreamOperator<T, T> {
+
+        @Override
+        public void processElement(StreamRecord<T> element) throws Exception {
+            throw new UnsupportedOperationException();
```

Comment:
I thought about it and agree that it is a good idea to use `output.collect(element)` here to simplify the additional behavior introduced by `BlockCheckpointBarrierOperator`.

As for the other `throw new UnsupportedOperationException();` in `EventReceivingOperator.processElement()`, I'll add more comments here to notify readers that this method should never be invoked, instead of leaving ambiguity that it might be expected for this method to be invoked and throw exceptions.

Reply:
Sounds good.